src/routes/slack.js

```diff
@@ -0,0 +1,183 @@
+/* @flow */
+
+const logger = require('../lib/logger');
+
+const rp = require('request-promise');
+const router = require('express').Router();
+
+const SlackCredentialModel = require('../models/SlackCredentialModel');
+const SlackCredential = require('../schemas/SlackCredential');
+
+const SongSearch = require('../models/SongSearch');
+const SongTemplate = require('../templates/slack/SongTemplate');
+
+const ArtistSearch = require('../models/ArtistSearch');
+const ArtistTemplate = require('../templates/slack/ArtistTemplate');
+
+const AlbumSearch = require('../models/AlbumSearch');
+const AlbumTemplate = require('../templates/slack/AlbumTemplate');
+
+const CardTemplate = require('../templates/slack/CardTemplate');
+const ResponseTemplate = require('../templates/slack/ResponseTemplate');
+
+const validParams = ['song', 'artist', 'album'];
+
+/**
+ * Middleware layer that verifies that both parameter and keyword
+ * are present in the command
+ * @type {void}
+ */
+router.post('/incoming', (req, res, next) => {
+  const { body } = req;
+  const { text, user_name } = body;
+
+  const tokens: [string] = text.split(' ');
+  const param: string = tokens[0];
+  const keyword: string = tokens.slice(1).join(' ');
+
+  if (validParams.indexOf(param) === -1 || keyword.length <= 0) {
+    res.status(200).send('Please enter a valid command!');
+    return;
+  }
+
+  req.user = user_name;
+  req.param = param;
+  req.keyword = keyword;
+  next();
+});
+
+/**
+ * Main router endpoint.
+ * Makes the call to the controller and returns the message
+ * @type {void}
+ */
+router.post('/incoming', (req, res) => {
+
+  const param: string = req.param;
+  const keyword: string = req.keyword;
+  const username: string = req.user;
+
+  let result: Promise<*>;
+  switch (param) {
+    case 'song':
+      const songSearch = new SongSearch(keyword, 5);
+      songSearch.search();
```

Comment:
There is a temporal coupling here I think.
If you don't call `songSearch.search();` first, the `songSearch.getResults()` will fail, right?
That's a code smell you can prevent I guess 🤔